### PR TITLE
Update due to keystone extrenal pkg move

### PR DESCRIPTION
**Pull request by stuggi**
update due to [1] which moves the content of the external pkg to the api sub module to remove the dependency on the keystone-operator internals and just have to import the api.

Also updates the import of lib-common to the current latest.

[1] https://github.com/openstack-k8s-operators/keystone-operator/pull/90

### CODE REVIEW

**Comment by gibizer** [comment]
The envtest fails as it does not find the keystone CRD any more via the api package. We might need to re-think how we get those CRDs in the test env now that the full package is not in our dependency but the CRD is only available in the full package. I will try to get back to this today to do a deeper look...

**Comment by stuggi** [comment]
> The envtest fails as it does not find the keystone CRD any more via the api package. We might need to re-think how we get those CRDs in the test env now that the full package is not in our dependency but the CRD is only available in the full package. I will try to get back to this today to do a deeper look...

yes, I have also started looking/thinking about it and testing right now:
- keep the list as we had it
- get the version of the api submodule we have in go.mod
- in envsetup install the module with version from above

**Comment by gibizer** [comment]
I think the main question we have to resolve is: How can we get the correct version of the CRD based on the version of the keystone/api module. I believe that now the keystone and the keystone/api modules are (can be) versioned separately. And now placement states correctly only the version of keystone/api it depends on. So from either:

-a) Deduce the version of the keystone module from the version of the keystone/api version in our go.mod and install that version of the keystone module in the test env to get the CRD. I'm not sure this is possible as there might not be 1:1 relationship between keystone/api and keystone module versions

- b) move the keystone CRD somehow to the keystone/api module so that we can get it from there. This would keep the API struct definitions and the generated CRD at the same module which is nice. But I'm not sure this can be done with operator-sdk.

-c) we take the API structs from the keystone/api module and generate the CRD locally. I have to look into how to trigger such regeneration.

**Comment by SeanMooney** [comment]
do me the CRDs are the public user facing API of the operator so it makes sense to me that they would be version with the go api and be part of the same module.

so if we can do b i would

**Comment by gibizer** [comment]
My only idea for #b) is to simlink the /config dir into /api/ . @stuggi, @dprince: could this be a way forward?

For #c) I  can generate the CRD locally from the API structs with:
``` bin/controller-gen crd paths=/home/gibi/go/pkg/mod/github.com/openstack-k8s-operators/keystone-operator/api@v0.0.0-20220913134025-daee844b58f7/v1beta1 output:crd:dir=/tmp/crds output:stdout``` We already generate the path to the locally downloaded keystone/api module. So we only need an extra step to run the generator. I will try to hack something up for it.

**Comment by SeanMooney** [comment]
i guess since the  crds are generated form the go code C is mostly the same a b so ya we could do that in the envtest case.

**Comment by stuggi** [comment]
I was testing a) but that is really hackish and if you want to use replace not easy to use. So I stopped it.

> My only idea for #b) is to simlink the /config dir into /api/ . @stuggi, @dprince: could this be a way forward?
could just symlink config/crd/base and could add a check in the `make manifests` that it is there.

> 
> For #c) I can generate the CRD locally from the API structs with: ` bin/controller-gen crd paths=/home/gibi/go/pkg/mod/github.com/openstack-k8s-operators/keystone-operator/api@v0.0.0-20220913134025-daee844b58f7/v1beta1 output:crd:dir=/tmp/crds output:stdout` We already generate the path to the locally downloaded keystone/api module. So we only need an extra step to run the generator. I will try to hack something up for it.

I think b) could work (have not tested) and might be the quickest. c) has the same end result, so could also do that.

**Comment by stuggi** [comment]
this is to demo the symlink approach. basically each operator need an update to move `/config/crd/bases` to `/api/bases` and create a link `/config/crd/bases` -> `/api/bases`. I also added a small enhancement so that replace is also checked in go.mod for a  module. Like in a commit of this PR I point to my private clone for keystone/mariadb where I did the above link change.

File: controllers/suite_test.go [hunk 1]
```diff
@@ -67,31 +67,46 @@ func TestAPIs(t *testing.T) {
...
 	for _, r := range f.Require {
 		if r.Mod.Path == moduleName {
-			return r.Mod.Version, nil
+			version = r.Mod.Version
+		}
+	}
+
+	// check for replacement config in go.mod for the named module
+	for _, r := range f.Replace {
+		if r.Old.Path == moduleName {
```
[the remark this answers is not in the document]
Comment (gibizer) [reply controllers/suite_test.go:92]: I'm mistaken. This now finds all the replacement and take the last one. Regardless it seems to be OK as there can only one such replacement for a given module

File: controllers/suite_test.go [hunk 2]
```diff
@@ -67,31 +67,46 @@ func TestAPIs(t *testing.T) {
...
+			name = r.New.Path
 		}
 	}
-	return "", fmt.Errorf("Cannot find %s in our go.mod file", moduleName)
 
+	if version != "" {
+		return name, version, nil
+	}
+
+	return name, "", fmt.Errorf("Cannot find %s in our go.mod file", moduleName)
```
Comment (gibizer) [inline controllers/suite_test.go:102]: Probably does not matter. I'm just curious what is the golang recommendation for returning an error with a partial result?

**Comment by gibizer** [comment]
> this is to demo the symlink approach. basically each operator need an update to move `/config/crd/bases` to `/api/bases` and create a link `/config/crd/bases` -> `/api/bases`. I also added a small enhancement so that replace is also checked in go.mod for a module. Like in a commit of this PR I point to my private clone for keystone/mariadb where I did the above link change.

If we can agree on this pattern for our operators then simlinking seems to be the clean approach. 
(Generating locally has the issue that I would need to run the generator from the test code to be able to ensure that the generator sees the correct dependency versions)

/lgtm

**Comment by stuggi** [comment]
> > this is to demo the symlink approach. basically each operator need an update to move `/config/crd/bases` to `/api/bases` and create a link `/config/crd/bases` -> `/api/bases`. I also added a small enhancement so that replace is also checked in go.mod for a module. Like in a commit of this PR I point to my private clone for keystone/mariadb where I did the above link change.
> 
> If we can agree on this pattern for our operators then simlinking seems to be the clean approach. (Generating locally has the issue that I would need to run the generator from the test code to be able to ensure that the generator sees the correct dependency versions)
> 
> /lgtm

@dprince what is your take on the symlink change in the operators?

File: controllers/suite_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (gibizer) [reply controllers/suite_test.go:92]: I think this is OK as is. I also believe that we cannot have two replacements for the same module

File: controllers/suite_test.go (hunk 2 above)
Comment (gibizer) [reply controllers/suite_test.go:102]: Thanks for the explanation.

**Comment by dprince** [comment]
My vote: store all the required CR's locally like we do for routev1: https://github.com/openstack-k8s-operators/placement-operator/blob/master/controllers/suite_test.go#L112

**Comment by stuggi** [comment]
> Is there a way to have the test pull the version of the keystone CRD from the go.mod version?

I had a quick look at that, but was not fully working. I'll revisit that again that we can compare.

**Comment by gibizer** [comment]
> Is there a way to have the test pull the version of the keystone CRD from the go.mod version?

The go.mod of placement only has the keystone-operator/api  not keystone-operator any more, so we can only pull keystone/api,  but that module does not have the CRD in it (hence the symlinking).

**Comment by stuggi** [comment]
> > Is there a way to have the test pull the version of the keystone CRD from the go.mod version?
> 
> The go.mod of placement only has the keystone-operator/api not keystone-operator any more, so we can only pull keystone/api, but that module does not have the CRD in it (hence the symlinking).

I have a working patch to load the base module using the api version details of the go.mod . will submit in a bit so that we can discuss that approach.

**Comment by stuggi** [comment]
> > > Is there a way to have the test pull the version of the keystone CRD from the go.mod version?
> > 
> > 
> > The go.mod of placement only has the keystone-operator/api not keystone-operator any more, so we can only pull keystone/api, but that module does not have the CRD in it (hence the symlinking).
> 
> I have a working patch to load the base module using the api version details of the go.mod . will submit in a bit so that we can discuss that approach.

@gibizer @dprince @SeanMooney please have a look at https://github.com/openstack-k8s-operators/placement-operator/pull/30

**Comment by gibizer** [comment]
> > > > Is there a way to have the test pull the version of the keystone CRD from the go.mod version?
> > > 
> > > 
> > > The go.mod of placement only has the keystone-operator/api not keystone-operator any more, so we can only pull keystone/api, but that module does not have the CRD in it (hence the symlinking).
> > 
> > 
> > I have a working patch to load the base module using the api version details of the go.mod . will submit in a bit so that we can discuss that approach.
> 
> @gibizer @dprince @SeanMooney please have a look at #30

I still think that storing (or symlinking) the CRD in the same module where the API Structs are stored is the cleanest approach. Those two things belong to the same API definition and there is a very high dependency between them.

**Comment by stuggi** [comment]
> > > > > Is there a way to have the test pull the version of the keystone CRD from the go.mod version?
> > > > 
> > > > 
> > > > The go.mod of placement only has the keystone-operator/api not keystone-operator any more, so we can only pull keystone/api, but that module does not have the CRD in it (hence the symlinking).
> > > 
> > > 
> > > I have a working patch to load the base module using the api version details of the go.mod . will submit in a bit so that we can discuss that approach.
> > 
> > 
> > @gibizer @dprince @SeanMooney please have a look at #30
> 
> I still think that storing (or symlinking) the CRD in the same module where the API Structs are stored is the cleanest approach. Those two things belong to the same API definition and there is a very high dependency between them.

yes I agree, storing in /api and add a symlink from the original place to be able to use the default create mechanism is the easiest and cleanest way. the other options PR #30 or store a copy of the required CRDs in the operator where it is needed, can lead to wrong versions if there is no corresponding base version to the sub module, or it got missed to sync them.

**Comment by stuggi** [comment]
I'll get a new one filed now that we have https://github.com/openstack-k8s-operators/keystone-operator/pull/102 and https://github.com/openstack-k8s-operators/mariadb-operator/pull/35

**Pull request closed without merging**